### PR TITLE
ENHANCEMENT: Debug class emits plain text for application/json requests

### PR DESCRIPTION
**Pull request by tractorcow**
Also see https://github.com/silverstripe/silverstripe-admin/pull/137

I've majorly cleaned up and refactored the html generation from Debug class into either CliDebugView or DebugView

### CODE REVIEW

**Comment by chillu** [comment]
Copying my comment to Damian from Slack: I’m not sure its the right level of abstraction (has nothing to do with Ajax), but then again that’s already a problem in core. I think when you requested application/json, plaintext errors are always better than HTML errors. But this might cause regressions for devs who relied on `isAjax()` for app logic

**Comment by dhensby** [comment]
Rather than saying that it's an "ajax" request - which is a bad way of determining how to respond IMO - we should have a better way of returning different response formats.

Can't the error handler look at the `Accept` header and respond appropriately instead of always producing HTML responses?

File: src/Control/HTTPRequest.php [hunk 1]
```diff
@@ -406,8 +406,11 @@ public function getURL($includeGetVars = false)
      */
     public function isAjax()
     {
+        $contentType = $this->getHeader('Content-Type') ?: $this->getHeader('content-type');
```
Comment (dhensby) [inline src/Control/HTTPRequest.php]: Shouldn't this be the `accept` header?

Better to use `$this->getAcceptMimetypes()`?

**Comment by tractorcow** [comment]
Request body can also have a content-type. :) If you are posting json_data this assumes ajax.

I wanted to use accept, but in my case the accept is `*/*` which is useless.

**Comment by tractorcow** [comment]
I'll see if I can fix graphql to send the correct accept header.

**Comment by tractorcow** [comment]
I've updated graphql to send the appropriate accept header, and updated the code to respect this now.

@dhensby how does this look now? :D

**Comment by tractorcow** [comment]
The critique has been raised that we probably shouldn't be using is_ajax in places where we currently do for certain purposes.

It may be better to forgo this fix in favour of solution-specific patches to each area.

The use case I had in mind was Debug::create_debug_view() correctly rendering text errors for ajax requests; However it should probably be comparing the `accept:` header in this case for text/html, and inverting the check.

**Comment by dhensby** [comment]
I think you're right - we're doing this backwards.

Better to check the Accept header for content types than infer format from request "type".

**Comment by tractorcow** [comment]
I will update this PR once I get more time... maybe later this week. :) /remindme 4 days

**Comment by tractorcow** [comment]
I've pushed up a new proof of concept... I need to write tests still before we can merge it.

**Comment by tractorcow** [comment]
Note: Still working on this, will hope to update shortly (this week, not today).

**Comment by tractorcow** [comment]
I'll need to update this, maybe beta2.

**Comment by tractorcow** [comment]
ok, looking to pick this up again. :)

**Comment by tractorcow** [comment]
@dhensby I've updated this PR with some cleanup of Debug HTML generation.

File: tests/php/Dev/DebugViewTest.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace SilverStripe\Dev\Tests;
+
+use SilverStripe\Dev\DebugView;
+use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Dev\Tests\DebugViewTest\ObjectWithDebug;
+
+class DebugViewTest extends SapphireTest
```
Comment (tractorcow) [inline tests/php/Dev/DebugViewTest.php]: Oh look a test for a class we've had 10 years.

File: src/Dev/CliDebugView.php [hunk 3]
```diff
@@ -129,4 +131,76 @@ public function renderVariable($val, $caller)
...
+     * Get debug text for this object
+     *
+     * @param mixed $val
+     * @return string
+     */
+    public function debugVariableText($val)
+    {
+        // Check debug
+        if (ClassInfo::hasMethod($val, 'debug')) {
+            return $val->debug();
```
Comment (dhensby) [inline src/Dev/CliDebugView.php]: A lot of the time these methods hard code html into the output - is there a way we could either strip it or pass a parameter to indicate we don't want html?

File: src/Dev/Debug.php [hunk 4]
```diff
@@ -88,118 +76,136 @@ public static function caller()
...
-    public static function endshow($val)
+    public static function endshow($val, $showHeader = true, HTTPRequest $request = null)
     {
-        if (!Director::isLive()) {
-            $caller = Debug::caller();
-            echo "<hr>\n<h3>Debug \n<span style=\"font-size: 65%\">($caller[class]$caller[type]$caller[function]()"
-                . " \nin " . basename($caller['file']) . ":$caller[line])</span>\n</h3>\n";
-            echo Debug::text($val);
-            die();
+        // Don't show on live
```
Comment (dhensby) [inline src/Dev/Debug.php]: This method could just be `static::show(); die;`?

**Comment by dhensby** [comment]
I like it, a couple of points made that could be ignored if you just want to get this over the line

File: src/Dev/Debug.php (hunk 4 above)
Comment (tractorcow) [inline src/Dev/Debug.php]: static::caller() doesn't work since it is another level deep.

File: src/Dev/CliDebugView.php (hunk 3 above)
Comment (tractorcow) [inline src/Dev/CliDebugView.php]: Unfortunately debug() doesn't allow context to be provided; We simply (and have to) assume that user-implemented debug() methods will output in the correct format.

User code can use Debug::create_debug_view() internally to achieve this, but we can't easily and probably shouldn't enforce it.

File: src/Dev/Debug.php (hunk 4 above)
Comment (dhensby) [inline src/Dev/Debug.php]: ah - ok

**Pull request merged**